contracts/external/cw-payroll/src/contract.rs

```diff
@@ -0,0 +1,589 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    from_binary, to_binary, Binary, Coin, CosmosMsg, Decimal, Deps, DepsMut, DistributionMsg, Env,
+    MessageInfo, Order, Response, StakingMsg, StdResult, Uint128,
+};
+use cw2::set_contract_version;
+use cw20::Cw20ReceiveMsg;
+use cw_denom::{CheckedDenom, UncheckedDenom};
+use cw_utils::{must_pay, nonpayable};
+
+use crate::error::ContractError;
+use crate::msg::{ExecuteMsg, InstantiateMsg, QueryMsg, ReceiveMsg};
+use crate::state::{
+    ValidatorRewards, VestingPayment, VestingPaymentStatus, STAKED_VESTING_BY_VALIDATOR,
+    VALIDATORS_REWARDS, VESTING_PAYMENT,
+};
+
+const CONTRACT_NAME: &str = "crates.io:cw-payroll";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    cw_ownable::initialize_owner(deps.storage, deps.api, msg.owner.as_deref())?;
+
+    // Check all params and create vesting payment with status based on denom
+    let vp = match msg.params.denom {
+        UncheckedDenom::Native(_) => {
+            // Validate all vesting payment params
+            let checked_params = msg.params.into_checked(deps.as_ref())?;
+
+            // Check amount sent matches amount in vesting payment
+            if checked_params.amount != must_pay(&info, &checked_params.denom.to_string())? {
+                return Err(ContractError::AmountDoesNotMatch);
+            }
+
+            // Create a new vesting payment
+            VestingPayment::new(checked_params)?
+        }
+        UncheckedDenom::Cw20(_) => {
+            // If configured for cw20, check no native funds included
+            nonpayable(&info)?;
+
+            // Validate all vesting payment params
+            let checked_params = msg.params.into_checked(deps.as_ref())?;
+
+            // Create new vesting payment
+            let mut vesting_payment = VestingPayment::new(checked_params)?;
+
+            // Set status to unfunded
+            vesting_payment.status = VestingPaymentStatus::Unfunded;
```

Comment:
small nit on design: I think it's a bit unintuitive that native vesting contracts can be funded immediately but cw20 payments need to be created and then funded in a separate step. I understand the technical reason for why this has to be, but an alternative design might be to just require both native and cw20 payments to be created and funded in two separate steps

Reply:
I think we can do it with the factory... will try that today.

Reply:
Done!